### ISSUE TITLE
Support disabling notifications

### ISSUE THREAD

**Issue by DanielleMaywood** [coder/coder#15513]
### Problem Description

Not all coder deployments will want notifications enabled, so it would be nice to be able to disable the feature. Currently, if you leave the notification settings as default, there are error logs that appear every time the system attempts to dispatch a notification.

### Desired Solution

Add a flag `CODER_NOTIFICATIONS_ENABLED` that defaults to `true` but can be set to `false` to disable notifications.

**Comment by bpmct** [comment]
My understanding is that if an SMTP server is not set, our server will throw warning logs, right?

Perhaps we should also disable notifications if the STMP server is not set. That way, a default Coder install doesn't have warning logs. I could be misunderstanding this though.

**Comment by DanielleMaywood** [comment]
The issue is that when we introduced notifications, defaults were set on the cli for configuring SMTP. So we can either retroactively drop these defaults and risk breaking existing customers installs or add this flag to allow people to disable the functionality entirely.

**Comment by bpmct** [comment]
What are the defaults that are set that would break customers?

**Comment by bpmct** [comment]
> Currently, we will always try to dispatch notifications. The only way for these defaults to not error is for there to be a local SMTP at localhost:587.

So perhaps we should change this default to "" and that is essentially the equivalent of us disabling SMTP notifications? I imagine this will break practically zero deployments and we can call it out in the changelog.

**Comment by dannykopping** [comment]
We should also keep in mind that webhook is also a delivery method.

I agree that _implicitly_ disabling notifications by checking if all the required configs are unset more ergonomic, but as Danielle correctly identified we'd have to enumerate which fields are required to make each delivery method work, which makes things complicated and less predictable for the operator.

Perhaps we can start by 1) adding the killswitch option and 2) only displaying the warning about misconfigured SMTP/webhook settings when the notification manager is instantiated (i.e. _once, rather than on every dispatch)?

IIRC I made it warn on each dispatch to make the likelihood of discovery by operators more likely, thinking that everyone would want to enable notifications, but given that we've had at least 1 customer now asking to disable them altogether it feels like we need an overall killswitch which is very clear in its usage.

**Comment by dannykopping** [comment]
I hear you about not showing warnings on a default install, and how we provide defaults so the installation works out-of-the-box. However, in those cases the functionality is non-optional whereas this very much is. Open SMTP servers exist but will largely be abused by spammers, so I don't think we can make this work out-of-the-box, at least for SMTP. We could default the installation to webhooks and have a loopback webhook receiver with some basic UI to display the received notifications in our dashboard, but that seems meh to me.

I think option 2 above is the most sensible. We'd have to call it out specifically in the release notes, but I doubt anyone is actually using the defaults as they are now to integrate with a real SMTP server.

---

As an aside, I think I recall how notifications got released without a killswitch. We had notifications behind an experiment (which effectively _was_ the killswitch) and when we removed the experiment it became enabled by default; this was an oversight on my part, apologies!
